### PR TITLE
OpenMPTarget hierarchical

### PR DESCRIPTION
**Pull request by rgayatri23**
A newer PR for Hierarchial parallelism in OpenMPTarget backend. 
Passes the first 11 incremental tests. 
Kokkos league level maps to teams distribute,
Thread level maps to parallel for and
Vector level maps to simd

Reduction only works for "+" operator.

### CODE REVIEW

**Comment by dalg24** [comment]
Add to whitelist

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp [hunk 1]
```diff
@@ -158,6 +595,9 @@ class OpenMPTargetExecTeamMember {
   KOKKOS_INLINE_FUNCTION int league_size() const { return m_league_size; }
   KOKKOS_INLINE_FUNCTION int team_rank() const { return m_team_rank; }
   KOKKOS_INLINE_FUNCTION int team_size() const { return m_team_size; }
+  KOKKOS_INLINE_FUNCTION void* impl_reduce_scratch() const {
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:598]: Does it need return a pointer to non-const?  This violate const-correctness in some way.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp [hunk 2]
```diff
@@ -572,9 +1037,8 @@ KOKKOS_INLINE_FUNCTION void parallel_for(
     const Impl::TeamThreadRangeBoundariesStruct<
         iType, Impl::OpenMPTargetExecTeamMember>& loop_boundaries,
     const Lambda& lambda) {
-  for (iType i = loop_boundaries.start; i < loop_boundaries.end;
-       i += loop_boundaries.increment)
-    lambda(i);
+#pragma omp for nowait schedule(static, 1)
+  for (iType i = loop_boundaries.start; i < loop_boundaries.end; i++) lambda(i);
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:1041]: why did you change the increment?

File: core/unit_test/incremental/Test10_HierarchicalBasics.hpp [hunk 3]
```diff
@@ -93,9 +94,17 @@ struct HierarchicalBasics {
 
 TEST(TEST_CATEGORY, IncrTest_10_Hierarchical_Basics) {
   HierarchicalBasics<TEST_EXECSPACE> test;
-  test.run(1, 4);
-  test.run(8, 16);
-  test.run(11, 13);
+
+  // OpenMPTarget backend only accepts >= 32 threads per team
```
[the remark this answers is not in the document]
Comment (rgayatri23) [reply core/unit_test/incremental/Test10_HierarchicalBasics.hpp:98]: No it does not need to be a multiple of 32. 
Actually I just realized that the clang compiler actually does (32+num_threads-requested-per-team) .
So essentially even if I assign team_size = 32, the actual block generated on the NVIDIA GPU is for 64 threads per team.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp [hunk 4]
```diff
@@ -50,6 +50,442 @@
 
 #include <Kokkos_Atomic.hpp>
 
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:54]: Yes sorry about this, even @crtrott pointed this out. I will be careful next time.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp (hunk 2 above)
Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:1041]: We only increment by 1 in every hierarchical level now. So the `increment` variable is no longer needed.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:598]: Yes: because the pointer will be converted to whatever the reduction type is and then needs to be written and read from.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp [hunk 5]
```diff
@@ -380,10 +822,15 @@ class TeamPolicyInternal<Kokkos::Experimental::OpenMPTarget, Properties...>
...
     m_league_size = league_size_request;
 
-    m_team_size = team_size_request;
+    // Minimum team size should be 32 for OpenMPTarget backend.
+    if (team_size_request < 32) {
+      printf(
+          "OpenMPTarget backend requires a minimum of 32 threads per team.\n");
+      exit(EXIT_FAILURE);
+    } else
+      m_team_size = team_size_request;
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:831]: should be abort.

**Review by dalg24 — APPROVED** [review]
